### PR TITLE
Feat/1248 search by ids : added an endpoint to query for ids

### PR DESCRIPTION
**Pull request by jberube**
# /searchByIds endpoint

Related to https://github.com/kids-first/kf-portal-ui/issues/1248

Added an endpoint to `kf-arranger` that does the search of participants by different ids.

## expected request

```
POST /searchByIds HTTP/1.1
Host: <REACT_APP_ARRANGER_API>
Content-Type: application/json
Authorization: Bearer <EGO TOKEN>

{
    "project": "2019_07_11_v1",
    "ids": [
        "PT_S5KPD5GG",
        "TARGET-30-PAPVRN",
        "DSD-IX-1218-0",
        "BS_V9FZVFQV",
        "BS_VT6C6N6K",
        "BS_T1SNKF50",
        "FM_0XEC843T",
        "FM_C9GV36EV"
    ]
}
```

## example result

``` json
{
    "participants": [
        {
            "search": "PT_S5KPD5GG",
            "type": "PARTICIPANT",
            "participantIds": [
                "PT_S5KPD5GG"
            ]
        },
        {
            "search": "TARGET-30-PAPVRN",
            "type": "PARTICIPANT",
            "participantIds": [
                "PT_FCPZ76X8"
            ]
        },
        {
            "search": "DSD-IX-1218-0",
            "type": "PARTICIPANT",
            "participantIds": [
                "PT_PNDXTWEV"
            ]
        },
        {
            "search": "BS_V9FZVFQV",
            "type": "BIOSPECIMEN",
            "participantIds": [
                "PT_1NZGDP3P"
            ]
        },
        {
            "search": "BS_VT6C6N6K",
            "type": "BIOSPECIMEN",
            "participantIds": [
                "PT_1NZGDP3P"
            ]
        },
        {
            "search": "BS_T1SNKF50",
            "type": "BIOSPECIMEN",
            "participantIds": [
                "PT_CYBYN18G"
            ]
        },
        {
            "search": "FM_0XEC843T",
            "type": "FAMILY",
            "participantIds": [
                "PT_4HP4EHRV",
                "PT_RZY9JG3M",
                "PT_A7W6NKPZ"
            ]
        },
        {
            "search": "FM_C9GV36EV",
            "type": "FAMILY",
            "participantIds": [
                "PT_PNDXTWEV",
                "PT_B8S85PP8",
                "PT_MPVVXY6J"
            ]
        },
        {
            "search": "DSD-IX-1218-0",
            "type": "SAMPLE EXTERNAL ID",
            "participantIds": [
                "PT_PNDXTWEV"
            ]
        }
    ]
}
```

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,33 +21,50 @@
   },
   "homepage": "https://github.com/kids-first/kf-arranger#readme",
   "dependencies": {
-    "@kfarranger/server": "^1.2.2",
+    "@kfarranger/server": "^1.2.3",
```
Comment (nyanofthemoon) [inline package.json]: 1.2.3 needed for this?

File: src/endpoints/searchByIds/bySampleExternalId.js [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  getSqon,
+  transform: (data, ids) => {
+    const participants = get(data, 'participant', [])
+      .filter(p => !!p);
+
+    return ids.map(id => {
+      const participantIds = participants
+        .filter(participant => {
+          const biospecimens = get(participant, 'biospecimens', []);
+          return biospecimens.some(bio => bio.external_sample_id === id);
```
Comment (nyanofthemoon) [inline src/endpoints/searchByIds/bySampleExternalId.js]: There can be more than one biospecimen with `bio.external_sample_id === id` ?

File: src/utils/normalizeResults.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    return normalizeResults(hits);
+  }
+
+  if (typeof results === 'object' && results !== null) {
+    return Object.keys(results)
+      .reduce((norm, key) => {
+        const prop = results[key];
+
+        // detect 'node' and shortcut them
+        if (key === 'node' && typeof prop === 'object' && prop !== null) {
```
Comment (nyanofthemoon) [inline src/utils/normalizeResults.js]: :)

File: src/utils/normalizeResults.test.js [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  it('shortcuts "node" property', () => {
+    expect(normalized.participant[0]).not.to.have.property('node');
+    expect(normalized.participant[0]).to.have.property('kf_id');
+  });
+
+  it('shortcuts "node" property recursively', () => {
+    expect(normalized.participant[0].biospecimens[0]).not.to.have.property('node');
+    expect(normalized.participant[0]).to.have.property('kf_id');
+  });
+});
```
Comment (nyanofthemoon) [inline src/utils/normalizeResults.test.js]: 🦄

File: package.json [hunk 5]
```diff
@@ -20,33 +21,50 @@
...
+    "@babel/core": "^7.5.5",
+    "@babel/plugin-proposal-object-rest-spread": "^7.5.5",
+    "@babel/plugin-proposal-optional-chaining": "^7.2.0",
+    "@babel/preset-env": "^7.5.5",
+    "@babel/register": "^7.5.5",
     "apollo-server-core": "1.3.2",
-    "babel-polyfill": "^6.26.0",
     "body-parser": "^1.18.2",
     "chalk-animation": "^1.5.0",
+    "core-js": "^3.1.4",
```
Comment (nyanofthemoon) [inline package.json]: Haven't used this one yet- hopefully it remains supported / open for a long time.

File: src/index.js [hunk 6]
```diff
@@ -1,4 +1,7 @@
-import 'babel-polyfill';
+// Necessary to polyfill Babel (replaces @babel/polyfill since 7.4.0)
+// see https://babeljs.io/docs/en/7.4.0/babel-polyfill
```
Comment (nyanofthemoon) [inline src/index.js]: Ah, ok. I understand.

File: src/endpoints/searchByIds/bySampleExternalId.js (hunk 2 above)
Comment (jberube) [reply src/endpoints/searchByIds/bySampleExternalId.js]: No, but `some` will stop iterating once one is found.

File: src/utils/normalizeResults.test.js (hunk 4 above)
Comment (jberube) [reply src/utils/normalizeResults.test.js]: Yes, :unicorn: !

File: package.json (hunk 5 above)
Comment (jberube) [reply package.json]: Replaces `babel-polyfill`.
see https://babeljs.io/docs/en/7.4.0/babel-polyfill

File: package.json (hunk 1 above)
Comment (jberube) [reply package.json]: I think so, though I can't remember why on top of my head.

**Pull request merged**